content/blog/single-message-transforms-swiss-army-knife-of-kafka-connect.asciidoc

```diff
@@ -0,0 +1,235 @@
+
+== Limitations
+
+As we've seen, single message transformations can help to address quite a few requirements that commonly come up for users of Kafka Connect.
+But there are limitations, too;
+Like MacGyver, who sometimes has to reach for some other tool than his beloved Swiss Army knife, you shouldn't think of SMTs as the perfect solution all the time.
+
+The biggest shortcoming is already hinted at in their name:
+SMTs only can be used to process _single_ records, one at a time.
+E.g. you cannot split up a record into multiple ones using an SMT, as they only can return (at most) one record.
+Also any kind of stateful processing, like aggregating data from multiple records, or correlating records from several topics is off limits for SMTs.
+For such use cases, you should be looking at stream processing technologies like https://kafka.apache.org/25/documentation/streams/developer-guide/[Kafka Streams] and https://flink.apache.org/[Apache Flink]; also integration technologies like https://camel.apache.org/[Apache Camel] can be of great use here.
+
+One thing to be aware of when working with SMTs is configuration complexity;
+when using generic, highly configurable SMTs, you might end up with lengthy configuration that's hard to grasp and debug.
+You might be better off implementing a bespoke SMT which is focussing on one particular task, leveraging the full capabilities of the Java programming language.
+
+[NOTE]
+.SMT Testing
+====
+Whether you use ready-made SMTs by means of configuration, or you implement custom SMTs in Java,
+testing your work is essential.
+
+While unit tests are a viable option for basic testing of bespoke SMT implementations,
+integration tests running against Kafka Connect connectors are recommended for testing SMT configurations.
+That way you'll be sure that the SMT can process actual messages and it has been configured the way you intended to.
+
+Testcontainers and the Debezium https://debezium.io/documentation/reference/integrations/testcontainers.html[support for Testcontainers] are a great foundation for setting up all the required components such as Apache Kafka, Kafka Connect, connectors and the SMTs to test.
+====
+
+A specific feature I wished for every now and then is the ability to apply SMTs only to a specific sub-set of the topics created or consumed by a connector.
+In particular if connectors create different kinds of topics (like an actual data topic and another one with with metadata),
+it can be desirable to apply SMTs only to the topics of one group but not the other.
+This requirement is captured in https://cwiki.apache.org/confluence/display/KAFKA/KIP-585%3A+Filter+and+Conditional+SMTs[KIP-585] ("Filter and Conditional SMTs"),
+please join the discussion on that one if you got requirements or feedback related to that.
+
+== Learning More
+
+There are several great presentations and blog posts out which describe in depth what SMTs,
```

Comment:
and blog post out "there" ... ?
which describe in depth what SMTs "are"?

Reply:
Uff, yes. Thanks!